### DISCUSSION TITLE
Woche 4

### DISCUSSION THREAD

**Discussion by Christian-2003** [Christian-2003/CrInGE#16] [Student Exchange]
Hallo zusammen,

in dieser Woche haben wir uns mit der :thinking: **Verhaltensmodellierung** unser :video_game: **Videospielengine** beschäftigt. Dazu haben wir mehrere UML-Aktivitätsdiagramme und UML-Sequenzdiagramme gezeichnet und in die :page_with_curl: [**Software Anforderungsspezifikation**](https://github.com/Christian-2003/CrInGE/blob/master/docs/SoftwareAnforderungsspezifikation.md) integriert.

###### Welteneditor

Ein Zentrales Element in der Videospielengine ist der :deciduous_tree: **Welteneditor**, mit welchem die Videospielwelten zusammengestellt werden können. Damit ihr euch das besser vorstellen könnt, haben wir hierzu ein UML-Aktivitätsdiagramm erstellt, welches das Verhalten des Welteneditors beschreibt:

![Aktivitätsdiagramm_Welteneditor](https://raw.githubusercontent.com/Christian-2003/CrInGE/b9e593cda683db5cd24059c9eb4c74179ca9020b/docs/resources/UML/Aktivit%C3%A4tsdiagramm_Welteneditor.drawio.png)

Allerdings ist es möglicherweise etwas kompliziert die Rolle und das Verhalten der :floppy_disk: **Assets** und des :mag_right: **Objektinspektors** zu verstehen. Daher haben wir für den Welteneditor zusätzlich ein UML-Sequenzdiagramm erstellt:

![Sequenzdiagramm_Welteneditor](https://github.com/Christian-2003/CrInGE/assets/80711177/f56846c3-8b16-47f5-8218-371c8408e006)

###### Programmierschnittstelle (API)

Um die Videospielwelt zu programmieren, bieten wir eine :computer: **API** an. Hierfür haben wir ebenfalls ein UML-Aktivitätsdiagramm gezeichnet:

![Aktivitätsdiagramm_API](https://github.com/Christian-2003/CrInGE/assets/80711177/db0ad282-a415-4cae-83e1-9898a775354e)

Dieses Diagramm beschreibt beispielhaft das Ändern von :key: **Attributen** für Objekte, die über den Welteneditor in die Videospielwelt eingefügt wurden. Aber auch hier haben wir für das bessere Verständnis zusätzlich noch ein UML-Sequenzdiagramm gezeichnet:

![Sequenzdiagramm_API](https://github.com/Christian-2003/CrInGE/assets/80711177/14987a74-a494-40c4-8fe5-11aa97dbd9bb)

###### Videospiel exportieren

Damit das entwickelte Videospiel von möglichst vielen :video_game: **Spielern** gespielt werden kann, gibt es selbstverständlich auch die Möglichkeit, das erstellte Videospiel zu :arrow_forward: **exportieren**. Hierzu gibt es ebenfalls ein UML-Aktivitätsdiagramm:

![Aktivitätsdiagramm_Exportieren](https://github.com/Christian-2003/CrInGE/assets/80711177/b80a7558-ea6c-4f18-9a69-2f690d7e42fc)

###### Weitere Aufgaben

Neben der Verhaltensmodellierung haben wir noch einige kleinere Aufgaben durchgeführt:

Unter Anderem haben wir die Dokumentation komplett einheitlich ins Deutsche übersetzt.

Des Weiteren haben wir eines der SOLID-Prinzipien ausgewählt, welches wir in unserer Implementierung anwenden möchten, ausgesucht.

Damit wäre auch schon das Ende des Blogs für diese Woche erreicht. Wir freuen uns auch euer Feedback und werden euch nächste Woche weitere Fortschritte berichten.

**Comment by lukpr0** [comment]
Hallo CrInGE-Team, diese Woche hätte ich ein paar Fragen.
1) Ihr schreibt an mehreren Stellen davon, das Attribute bearbeitet werden sollen. Mir ist noch nicht ganz klar, was für Attribute gemeint sind, könnt ihr Beispiele nennen?

2) Verstehe ich nicht, wie das 2. Sequenzdiagramm zu lesen ist. Was ist z.B. gemeint mit Instanz der Objektklasse erstellen? Warum muss dazu überprüft werden, ob ein Objekt existiert?

**Reply by Christian-2003** [reply]
Hallo Lukas,

bezüglich deiner Fragen:

1. Alle Objekte in der Videospielwelt besitzen Attribute, wie beispielsweise deren Farbe, Textur, Position, Kollisionsverhalten oder Ähnliches. Solche Attribute können dann über die API zur Laufzeit bearbeitet werden.
2. Beim zweiten Sequenzdiagramm stellt die API eine Klasse zur Verfügung, über welche auf die Videospielobjekte in der Videospielwelt zugegriffen werden kann. Wie genau diese Videospielobjekte über die API referenziert werden (bspw. IDs) ist noch unklar. Unabhängig davon muss aber auf jeden Fall überprüft werden, ob das referenzierte Objekt existiert, bevor eine Instanz der Klasse der API erstellt werden kann.

Ich hoffe, ich konnte deine Fragen beantworten.

**Comment by hmi95** [comment]
Frage zum Aktivitätsdiagramm für den Welteneditor: Angenommen ich möchte als Benutzer an einem Objekt in der Welt drei Attribute verändern. Muss ich den Ablauf "Objekt in Welt auswählen" > "Zu Änderndes Atrribut auswählen" > "Neuen Attributwert eingeben" > "Neuen Attributwert übernehmen" > Objekt speichern" dreimal durchlaufen? 
![grafik](https://github.com/Christian-2003/CrInGE/assets/84066957/270f78fa-322c-49de-9ba5-4a5b91c4dbd1)

**Reply by TimSchnur** [reply]
Hallo Herr Ichters,
da ist uns ein kleiner Fehler im UML unterlaufen.
Es ist natürlich so gedacht, dass man mehrere Attributwerte abändern kann, bevor man das Objekt speichert, auch ohne es mehrmals auswählen zu müssen. Wir haben das UML bereits überarbeitet und im Blog aktualisiert.
